tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+	tostring = tostring,
+	select   = select,
+
+	-- luatex specific libraries
+	lfs      = {isfile=lfs.isfile},
+	kpse     = kpse,
+
+	-- memoize-extract specific global
+	STAGE    = STAGE,
+}
+
+local exit
+if STAGE == "testing" then
+	-- in testing environment avoid exiting the whole test
+	-- -> instead raise an error which can be catched
+
+	-- store the error function independent of the environment
+	local error = error
+	exit = {
+		error = function() error("exited with error") end,
+		warn  = function() error("exited with warn") end,
+		succ  = function() error("exited with succ") end,
+	}
+else
+	local os_exit = os.exit
+	exit = {
+		error = function() os_exit(11) end,
+		warn  = function() os_exit(10) end,
+		succ  = function() os_exit(0) end,
+	}
+end
+
+do
+	-- I don't like using the debug library, but getting a traceback here is a
+	-- must to find where the error originates from
+	local debug_traceback = debug.traceback
+	-- use the lua error function -> exits immediately
+	local error = error
+
+	-- Prevent trying to change the environment.
+	local function bad_index(...)
+		local msg = "Attempt to access an undefined index:"
+		for i = 2, select("#", ...) do
+			msg = msg ..tostring(select(i, ...)).." "
+		end
+		msg = msg.."\n\n"..debug_traceback(nil, 2)
+		error(msg)
+	end
+	setmetatable(env, {
+		__index     = bad_index,
+		__metatable = false,
+		__newindex  = bad_index,
+	})
+end
+
+_ENV = env
+----------------------------------
+-- restricted area startes here --
+----------------------------------
+
+---------------
+-- pathutils --
+---------------
+-- -> probably moved to a different library eventually
+local pathlib = {}
+do
+	-- other projects like penlight or l3build do stuff with different pathseps, according to
+	-- https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalize-separators
+	-- and
+	-- https://retrocomputing.stackexchange.com/questions/28344/since-when-does-windows-support-forward-slash-as-path-separator
+	-- windows since quite a while also works with / as pathsep.
+	-- Thus, this pathlib will normalize paths for having / as pathsep before working on paths
+	if os.type == "windows" then
+		---Normalize path such that windows also uses /
+		---@param path string
+		---@return string
+		function pathlib.path_normalize(path)
+			return path:gsub("\\", "/")
+		end
+	else
+		---Unix already uses / as pathsep -> no-op
+		---@param path string
+		---@return string
+		function pathlib.path_normalize(path)
+			return path
+		end
+	end
+
+	-- still windows paths work with disk specifiers -> special handling required
+	if os.type == "windows" then
+		---Check if path is an absolute path
+		---@param path string
+		---@return boolean? is_abs
+		---@return string? err_msg
+		function pathlib.path_is_absolute(path)
+			local err
+			path, err = pathlib.sanitize_path(path)
+			if not path then return nil, err end
+
+			return path:sub(2,2) == ":" and path:sub(3,3) == "/"
+		end
```

Comment:
[Windows paths are complicated.](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats) `\\server\share\path\to\file`, `\\.\C:\path\to\file`, and `\\?\C:\path\to\file` are all technically absolute paths. You don't currently support those path formats, and I see no reason for you to support them, but I'd recommend adding a comment stating `UNC paths aren't supported`.

Reply:
Oh yes true. Thanks for the note :+1: (I just added that comment in 2111415)

Reply:
Oh yes true. Thanks for the note 👍 (I just added that comment in https://github.com/atticus-sullivan/memoize/commit/21114155669ff042d3ac3c0dd3ddeddf879b9c2f)